aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,736 @@
+    /// transaction).
+    pub fn into_change_set(self) -> AggregatorChangeSet {
+        let NativeAggregatorContext {
+            aggregator_data, ..
+        } = self;
+        let AggregatorData {
+            destroyed_aggregators,
+            aggregators,
+            ..
+        } = aggregator_data.into_inner();
+
+        let mut changes = BTreeMap::new();
+
+        // First, process all writes and deltas.
+        for (id, aggregator) in aggregators {
+            let Aggregator {
+                value,
+                state,
+                limit,
+            } = aggregator;
+
+            let change = match state {
+                AggregatorState::Data => AggregatorChange::Write(value),
+                AggregatorState::PositiveDelta => {
+                    let delta_op = DeltaOp::Addition { value, limit };
+                    AggregatorChange::Merge(delta_op)
+                }
+            };
+            changes.insert(id, change);
+        }
+
+        // Additionally, do not forget to delete destroyed values from storage.
+        for id in destroyed_aggregators {
+            changes.insert(id, AggregatorChange::Delete);
+        }
+
+        AggregatorChangeSet { changes }
+    }
+}
+
+// ================================= Natives =================================
+
+/// All aggregator native functions. For more details, refer to code in
+/// `aggregator_factory.move` and `aggregator.move`.
+pub fn aggregator_natives(aggregator_addr: AccountAddress) -> NativeFunctionTable {
+    native_functions::make_table(
+        aggregator_addr,
+        &[
+            ("aggregator", "add", Arc::new(native_add)),
+            ("aggregator", "read", Arc::new(native_read)),
+            ("aggregator", "destroy", Arc::new(native_destroy)),
+            ("aggregator", "sub", Arc::new(native_sub)),
+            (
+                "aggregator_factory",
+                "new_aggregator",
+                Arc::new(native_new_aggregator),
+            ),
+        ],
+    )
+}
+
+/// Move signature:
+/// fun new_aggregator(
+///   aggregator_factory: &mut AggregatorFactory,
+///   limit: u128
+/// ): Aggregator;
+fn native_new_aggregator(
+    context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    if !cfg!(any(test, feature = "aggregator-extension")) {
+        return Err(not_supported_error());
+    }
+    assert!(args.len() == 2);
+
+    // Extract fields: `limit` of the new aggregator and a `phantom_handle` of
+    // the parent factory.
+    let limit = pop_arg!(args, u128);
+    let handle = get_handle(&pop_arg!(args, StructRef))?;
+
+    // Get the current aggregator data.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    // Every aggregator instance uses a unique key in its id. Here we can reuse
```

Comment:
let's separate and test key generation logic (against separately computed what we should expect).
Additionally, let's make sure to confirm with any updates that happened to table handles (u256 discussion)

Reply:
I ll make a separate PR for that